File: 0_1.baekjoon/graph/java/src/BOJ17836.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class BOJ17836 {
    private static final int PATH = 0;
    private static final int WALL = 1;
    private static final int GRAM = 2;

    private static int N, M, T;
    private static int[][] board;
    private static int[][] d = {{1, 0}, {0, 1}, {0, -1}, {-1, 0}};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        T = Integer.parseInt(st.nextToken());

        board = new int[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int answer =  bfs();
        System.out.println(answer == -1 ? "Fail" : answer);
    }

    private static int bfs() {
        boolean[][][] visited = new boolean[N][M][2];
        Queue<int[]> q = new LinkedList<>();
        visited[0][0][0] = true;
        q.add(new int[] {0, 0, 0, 0});

        while (hasElement(q)) {
            int row = q.peek()[0];
            int col = q.peek()[1];
            int gram = q.peek()[2];
            int spend = q.poll()[3];

            if (row == N - 1 && col == M - 1) {
                return spend <= T ? spend : -1;
            }

            for (int i = 0; i < 4; i++) {
                int nr = row + d[i][0], nc = col + d[i][1];

                if (nr < 0 || nr >= N || nc < 0 || nc >= M) continue;

                if (visited[nr][nc][gram]) continue;

                if (board[nr][nc] == WALL) {
                    if (gram == 1) {
                        visited[nr][nc][gram] = true;
                        q.add(new int[]{nr, nc, gram, spend + 1});
                    }
                } else if (board[nr][nc] == GRAM) {
                    visited[nr][nc][1] = true;
                    q.add(new int[]{nr, nc, 1, spend + 1});
                } else {
                    visited[nr][nc][gram] = true;
                    q.add(new int[]{nr, nc, gram, spend + 1});
                }

            }
        }

        return -1;
    }

    private static boolean hasElement(Queue<int[]> q) {
        return !q.isEmpty();
    }
}

File: 0_1.baekjoon/graph/kotlin/src/BOJ17836.kt
import java.util.*

private val PATH = 0
private val WALL = 1
private val GRAM = 2
private val d = arrayOf(arrayOf(1, 0), arrayOf(0, 1), arrayOf(-1, 0), arrayOf(0, -1))

private var N = 0
private var M = 0
private var T = 0
private lateinit var board: Array<Array<Int>>
fun main() = with(System.`in`.bufferedReader()) {
    val(n, m, t) = readLine().split(" ").map { it.toInt() }
    N = n
    M = m
    T = t

    board = Array(n) {
        readLine().split(" ").map { it.toInt() }.toTypedArray()
    }
    val answer = bfs()
    print(if (answer == -1) "Fail" else answer)
}

private fun bfs(): Int {
    val visited = Array(N){Array(M) {Array(2) { false} } }
    val q: Queue<Array<Int>> = LinkedList()

    q.add(arrayOf(0, 0, 0, 0))
    visited[0][0][0] = true

    while (q.isNotEmpty()) {
        val (row, col, gram, spend) = q.poll()

        if (row == N - 1 && col == M - 1) {
            return if (spend <= T) spend else -1
        }

        for (i in d.indices) {
            val nr = row + d[i][0]
            val nc = col + d[i][1]

            if (nr < 0 || nr >= N || nc < 0 || nc >= M) continue

            if (visited[nr][nc][gram]) continue

            if (board[nr][nc] == WALL) {
                if (gram == 1) {
                    visited[nr][nc][gram] = true
                    q.add(arrayOf(nr, nc, gram, spend + 1))
                }
            } else if (board[nr][nc] == GRAM) {
                visited[nr][nc][1] = true
                q.add(arrayOf(nr, nc, 1, spend + 1))
            } else {
                visited[nr][nc][gram] = true
                q.add(arrayOf(nr, nc, gram, spend + 1))
            }
        }

    }

    return -1
}
